R/object.R

```diff
@@ -373,3 +373,141 @@ Object.get_or_upload_file <- function(local_path, vault_id, vault_path, filename
+            # Convert JSON string to an R structure
+            filters <- jsonlite::fromJSON(filters,
+                                          simplifyVector = FALSE,
+                                          simplifyDataFrame = TRUE,
+                                          simplifyMatrix = FALSE)
+        }
+        # Add filters to request body
+        body = modifyList(body, list(filters=filters))
+    }
+
+    path <- paste("v2/objects", paste(id), "data", sep="/")
+
+    tryCatch({
+        res <- .request('POST', path=path, body=body, env=env)
+        return(formatSolveBioQueryResponse(res, row.names=row.names))
+    }, error = function(e) {
+        cat(sprintf("Query failed: %s\n", e$message))
+    })
+}
+
+
+#' Object.query
+#'
+#' Queries a SolveBio file (vault object) and returns an R data frame containing all records.
+#' Returns a single page of results otherwise (default).
+#'
+#' @param id The ID of a SolveBio file (vault object).
+#' @param paginate When set to TRUE, retrieves all records (memory permitting).
+#' @param env (optional) Custom client environment.
+#' @param ... (optional) Additional query parameters (e.g. filters, limit, offset).
+#'
+#' @examples \dontrun{
+#' Object.query("12345678790", paginate=TRUE)
+#' }
+#'
+#' @references
+#' \url{https://docs.solvebio.com/}
+#'
+#' @export
+Object.query <- function(id, paginate=FALSE, env = solvebio:::.solveEnv, ...) {
+    params <- list(...)
+    params$id <- id
+    params$env <- env
+
+    # Retrieve the list of fields
+    # NOTE: There is no inherent order to these fields, unless the file is a TSV or CSV.
+    params$row.names <- do.call(Object.fields, list(id))$fields
+
+    # Retrieve the first page of results
+    response <- do.call(Object.data, params)
+    df <- response$result
+    offset <- response$offset
+    total <- response$total
+
+    # Continue to make requests for data if pagination is enabled and there are more records
+    while (isTRUE(paginate) && !is.null(offset)) {
+        params$offset <- offset
+        response <- do.call(Object.data, params)
+        df_page <- response$results
+        df <- dplyr::bind_rows(df, df_page)
+        offset <- response$offset
+    }
+
+    if (!isTRUE(paginate) && is.null(total)) {
+        warning(paste("This call returned only the first page of records. To retrieve more pages automatically,",
+                      "please set paginate=TRUE when calling Object.query().", call. = FALSE))
+    }
+
+    return(df)
+}
+
+
+#' Object.fields
+#'
+#' Retrieves the list of fields for a file (JSON, CSV, or TSV).
+#'
+#' @param id The ID of a SolveBio file (vault object).
+#' @param env (optional) Custom client environment.
+#' @param ... (optional) Additional query parameters (e.g. limit, offset).
+#'
+#' @examples \dontrun{
+#' Object.fields("1234567890")
+#' }
+#'
+#' @references
+#' \url{https://docs.solvebio.com/}
+#'
+#' @export
+Object.fields <- function(id, env = solvebio:::.solveEnv, ...) {
```

Comment:
@nikolamaric do you think we should return the response (as currently works) or return the contents of `$fields`?

Reply:
In the python client we return list of field, maybe we can do the same here too?

Reply:
Do you want to try to make that change and test it? The docs would need an update for that fields example.

Reply:
Sure. Will do both tomorrow.